src/Libraries/Microsoft.Extensions.Telemetry/Buffering/GlobalBuffer.cs

```diff
@@ -0,0 +1,189 @@
+        TimeProvider timeProvider)
+    {
+        _options = Throw.IfNull(options);
+        _timeProvider = timeProvider;
+        _bufferedLogger = bufferedLogger;
+        _category = Throw.IfNullOrEmpty(category);
+        _ruleSelector = Throw.IfNull(ruleSelector);
+        _lastKnownGoodFilterRules = LogBufferingFilterRuleSelector.SelectByCategory(_options.CurrentValue.Rules.ToArray(), _category);
+        _optionsChangeTokenRegistration = options.OnChange(OnOptionsChanged);
+    }
+
+    public void Dispose()
+    {
+        if (_disposed)
+        {
+            return;
+        }
+
+        _disposed = true;
+
+        _optionsChangeTokenRegistration?.Dispose();
+    }
+
+    public bool TryEnqueue<TState>(LogEntry<TState> logEntry)
+    {
+        if (_timeProvider.GetUtcNow() < _lastFlushTimestamp + _options.CurrentValue.AutoFlushDuration)
+        {
+            return false;
+        }
+
+        IReadOnlyList<KeyValuePair<string, object?>>? attributes = logEntry.State as IReadOnlyList<KeyValuePair<string, object?>>;
+        if (attributes is null)
+        {
+            // we expect state to be either ModernTagJoiner or LegacyTagJoiner
+            // which both implement IReadOnlyList<KeyValuePair<string, object?>>
+            // and if not, we throw an exception
+            Throw.InvalidOperationException(
+                $"Unsupported type of log state detected: {typeof(TState)}, expected IReadOnlyList<KeyValuePair<string, object?>>");
+        }
+
+        if (_ruleSelector.Select(_lastKnownGoodFilterRules, logEntry.LogLevel, logEntry.EventId, attributes) is null)
+        {
+            // buffering is not enabled for this log entry,
+            // return false to indicate that the log entry should be logged normally.
+            return false;
+        }
+
+        SerializedLogRecord serializedLogRecord = SerializedLogRecordFactory.Create(
+            logEntry.LogLevel,
+            logEntry.EventId,
+            _timeProvider.GetUtcNow(),
+            attributes,
+            logEntry.Exception,
+            logEntry.Formatter(logEntry.State, logEntry.Exception));
+
+        if (serializedLogRecord.SizeInBytes > _options.CurrentValue.MaxLogRecordSizeInBytes)
+        {
+            SerializedLogRecordFactory.Return(serializedLogRecord);
+            return false;
+        }
+
+        lock (_bufferSwapLock)
+        {
+            _buffer.Enqueue(serializedLogRecord);
+        }
+
+        _ = Interlocked.Add(ref _activeBufferSize, serializedLogRecord.SizeInBytes);
+        TrimExcessRecords();
+
+        return true;
+    }
+
+    public void Flush()
+    {
+        _lastFlushTimestamp = _timeProvider.GetUtcNow();
+        ConcurrentQueue<SerializedLogRecord> bufferToFlush;
+        lock (_bufferSwapLock)
+        {
+            bufferToFlush = _buffer;
+            _buffer = new ConcurrentQueue<SerializedLogRecord>();
+        }
+
+        Interlocked.Exchange(ref _activeBufferSize, 0);
```

Comment:
This must be in the lock above, otherwise this could be throwing away some amount.